Mathlib/CategoryTheory/Comma/Basic.lean

```diff
@@ -327,23 +367,25 @@ def preLeft (F : C ⥤ A) (L : A ⥤ T) (R : B ⥤ T) : Comma (F ⋙ L) R ⥤ Co
       w := by simpa using f.w }
 #align category_theory.comma.pre_left CategoryTheory.Comma.preLeft
 
-instance (F : C ⥤ A) (L : A ⥤ T) (R : B ⥤ T) [F.Faithful] : (preLeft F L R).Faithful where
-  map_injective {X Y} f g h := hom_ext _ _ (F.map_injective (congrArg CommaMorphism.left h))
-    (by apply congrArg CommaMorphism.right h)
+/-- `Comma.preLeft` is a particular case of `Comma.map`,
+but with better definitional properties. -/
+def preLeftIso (F : C ⥤ A) (L : A ⥤ T) (R : B ⥤ T) :
+    preLeft F L R ≅ map (F ⋙ L).rightUnitor.inv (R.rightUnitor.hom ≫ R.leftUnitor.inv) :=
+  NatIso.ofComponents (fun X => isoMk (Iso.refl _) (Iso.refl _))
+
+instance (F : C ⥤ A) (L : A ⥤ T) (R : B ⥤ T) [F.Faithful] : (preLeft F L R).Faithful :=
+  Functor.Faithful.of_iso (preLeftIso F L R).symm
```

Comment:
Here we use `preLeftIso`. If typeclass search calls unification, then it only allows unfolding of declarations that are `reducible` or projections on instances classes. Thus `preLeftIso` will not unfold and unification might fail because of this. 

This is a general policy (see the corresponding library note). While things build right now, we need to make sure theoretically defeq instances are functionally defeq to avoid future problems.

Reply:
Does this explanation help?

Reply:
`Functor.Faithful` is a Prop, so that I cannot see how this could be a problem. For `Functor.Full`, it may, but then I do not see how making `preLeftIso` reducible would improve things: the definitions `Full.ofIso`, `NatIso.ofComponents` and `Iso.refl` are not reducible either... (I think that `Full` and `IsEquivalence` should be made Props, because we should not rely on the definitional properties of the witness given by a `Full` instance.)

Reply:
Yes. This is bad example. I just took the first one. Apologies. 

All of `Full.ofIso` etc. should be made reducible (though not in this PR). Or there should be some refactor to carry no data. 

As I mentioned, I realize there is already a problem with the library as it currently exists. I just don't want to add to that problem here and make more to fix later.

Reply:
I think the better plan is to make `Full` and `IsEquivalence` a `Prop` in a near future. Then, the content of this PR would not require any modification.

Reply:
I think that is good idea yet. But I don't understand the force of the objection to marking things `reducible` for now, tagging with an issue to remove data from `Full` and `IsEquivalence`, and then cleaning things up later. 

Could you explain a bit more? Thanks

Reply:
I guess it is just more work to do when doing the refactor, without much gain (since `Functor.Full.of_iso` is not marked `reducible`) ?

Reply:
Is `Full.ofIso` used before this in instance constructions? Either way, it should be made `reducible` also since it used here.

Reply:
We really need to get the notes back into the API documentation: https://leanprover-community.github.io/mathlib_docs/notes.html#reducible%20non-instances

Reply:
There are other instances which uses `Full.ofIso`... I understand that there might be a problem here.

However, the right solution (*which I intend to implement*) is obviously to remove data from `Full` and `IsEquivalence` (and there are even more problematic instances: basically all the `PreservesLimit` stuff... nothing is reducible there). For instances where the carried data is of an extreme importance (algebraic operations on types, preorders, etc), I understand the need to mark things as reducible, but this is not at all the case here! Even if we were to mark everything as reducible, definitions of instances may be more automatically unfolded, but it is mostly unlikely that we get defeq of two concurrent instances (because in category theory definitions, we have extra compositions with identities everywhere and compositions may be "associated" in differents ways). As Andrew points out, if I were to mark a few definitions as reducible now, it would be more work afterwards when the `Prop`-refactor is done...

The `reducible` attribute is almost never used in the category theory API. Maybe we should use it more: one annoying thing is that when we define the `Hom` type of a category and setup an `ext` lemma for this `Hom` type, then we have to systematically restate the ext lemma for the type synonym of morphisms `X ⟶ Y` once the `Category` instance has been defined... I am not qualified to know how this could be fixed by marking certains things reducible, but this is a real issue...